plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/n4js.xsemantics

```diff
@@ -958,18 +958,25 @@ from {
 		if (lunknown && runknown) {
 			T = r
 		} else {
-			val lnum = l.declaredType == G.booleanType || G.isNumeric(l.declaredType);
-			val rnum = r.declaredType == G.booleanType || G.isNumeric(r.declaredType);
-			val undef =  l.declaredType == G.undefinedType  || l.declaredType == G.nullType
-					|| r.declaredType == G.undefinedType || r.declaredType == G.nullType;
-
-			if ((lnum && rnum) || ((undef && (lnum || rnum)))) {
+			val lnum = G.isNumericOperand(l);
+			val rnum = G.isNumericOperand(r);
+			
+			if (lnum && rnum) {
 				T = G.numberTypeRef
 			} else {
-				if ( (lunknown || runknown) && (lnum || rnum || undef)  ) {
-					T = if (lunknown) l else r
+				if ( (lunknown || runknown) && (lnum || rnum)  ) {
+					// one is numeric, the other unknown --> 
+					T = typeSystemHelper.createUnionType(G, G.numberTypeRef, G.stringTypeRef);
 				} else {
-					T = G.stringTypeRef
+					val lMayNum = lnum || G.containsNumericOperand(l);
+					val rMayNum = rnum || G.containsNumericOperand(r);
```

Comment:
`lMayNum`/`rMayNum` must be true also if `l`/`r` is a super type of `number` (i.e. `any`).

At the moment, the following code breaks:
```
let a: any = 5;
let str: string = 40 + a;  // does not show error
str.trim();  // exception at runtime
```

Reply:
Good catch. Added

```
// XPECT type of a+n --> union{number,string}
a+n
// XPECT type of a+s --> string
a+s
```